lib/logflare/partners.ex

```diff
@@ -0,0 +1,103 @@
+defmodule Logflare.Partners do
+  @moduledoc """
+  Handles Partner context
+  """
+  import Ecto.Query
+
+  alias Logflare.Partners.Partner
+  alias Logflare.Repo
+  alias Logflare.Users
+
+  @spec new(binary()) :: {:ok, Partner.t()} | {:error, any()}
+  @doc """
+  Creates a new partner with given name and token to be encrypted
+  """
+  def new(name) do
+    %Partner{}
+    |> Partner.changeset(%{name: name})
+    |> Repo.insert()
+  end
+
+  @spec list() :: [Partner.t()]
+  @doc """
+  Lists all partners
+  """
+  def list(), do: Repo.all(Partner)
+
+  @spec list_users(Partner.t()) :: [User.t()]
+  @doc """
+  Lists all users created by a partner
+  """
+  def list_users(%Partner{token: token}) do
+    query =
+      from(p in Partner,
+        join: u in assoc(p, :users),
+        where: p.token == ^token,
+        select: u
+      )
+
+    Repo.all(query)
+  end
+
+  @spec get_by_token(binary()) :: Partner.t() | nil
+
+  @doc """
+  Fetch single entry by given token
+  """
+  def get_by_token(token), do: Repo.get_by(Partner, token: token)
+
+  @spec create_user(Partner.t(), map()) ::
+          {:ok, %{user: User.t(), user_token: binary()}} | {:error, any()}
+  @doc """
+  Creates a new user and associates it with given partner
+  """
+  def create_user(%Partner{} = partner, params) do
+    user_token = Ecto.UUID.generate()
+
+    params =
+      Map.merge(params, %{
+        "provider_uid" => Ecto.UUID.generate(),
+        "provider" => "email",
+        "token" => user_token
+      })
```

Comment:
these autogen values should be in the changeset

Reply:
Would need to create a new function in `Users.insert_user` to handle it. Can do it if you think it's worth it